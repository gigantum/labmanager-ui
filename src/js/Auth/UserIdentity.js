//vendor
import {
  graphql,
} from 'react-relay'
//mutations
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation'
//environment
import {fetchQuery} from 'JS/createRelayEnvironment';

const userIdentityQuery = graphql`
  query UserIdentityQuery{
    userIdentity{
      id
      username
      email
      givenName
      familyName
      isSessionValid
    }
  }
`;

const UserIdentity = {
  getUserIdentity: () => {

    return new Promise((resolve, reject) =>{

      let fetchData = function(){

        fetchQuery(userIdentityQuery(), {}).then((response, error) => {

          if(response){
            resolve(response)
          }else{

            reject(response)
          }
        }).catch((error) =>{

          reject(error)
        })
      }

      fetchData()
    })
  },
  removeUserIdentity: () => {
    RemoveUserIdentityMutation(()=>{
      console.log('removed user')
    })
  }

}

export default UserIdentity
