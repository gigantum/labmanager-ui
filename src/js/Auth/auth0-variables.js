export const AUTH_CONFIG = {
  domain: 'calum.auth0.com',
  clientId: '9wFN5CZA4Z9xa1r8KZIZbevpjHG7bAab',
  callbackUrl: window.location.origin + '/callback'
}
