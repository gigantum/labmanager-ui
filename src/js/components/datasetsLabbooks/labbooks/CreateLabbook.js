import React from 'react'
import { QueryRenderer, graphql } from 'react-relay'

import CreateLabbookMutation from '../../../mutations/CreateLabbookMutation'

// const CreatePageViewerQuery = graphql`
//   query CreateLabbookQuery {
//     viewer {
//       id
//     }
//   }
// `;

export default class CreateLabbook extends React.Component {
  constructor(props){
  	super(props.props);
    console.log(props)
  	this.state = {'modal_visible': false};
  }
  _createLabbook(){
    let description = 'describe it';
    let name = 'tempLabbook3'// + Math.floor((Math.random()*1000));
    let viewerId = 'calum';
    console.log(this)
    //CreateLabbookMutation(description, name, viewerId,  () => this.props.history.replace(`/home`))
  }
  _showModal(){
    this.setState({'modal_visible': true})
  }
  _hideModal(){
    this.setState({'modal_visible': false})
  }
  render(){

    return(
        <div className='create-labbook__container'>
            <div className={!this.state.modal_visible ? 'create-labbook__modal hidden' : 'create-labbook__modal'}>
              <div className='create-labbook__modal-inner-container flex flex-column justify--space-around'>
                <div>
                  <label>Name</label><input type='text'></input>
                </div>
                <div>
                  <label>Description</label><input type='text'></input>
                </div>
                <div>
                  <button className='pa3 bg-black-10 bn dim ttu pointer' onClick={() => this._createLabbook()}>
                  Create Labbook</button>
                </div>
              </div>
            </div>

            <button className='pa3 bg-black-10 bn dim ttu pointer' onClick={() => this._showModal()}>Create Labbook</button>
        </div>
      )
  }
}
