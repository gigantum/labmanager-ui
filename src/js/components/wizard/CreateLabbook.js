//vendor
import React from 'react'
import uuidv4 from 'uuid/v4'
//utilities
import validation from 'JS/utils/Validation'
//components
import LoginPrompt from 'Components/labbook/branchMenu/LoginPrompt'
//mutations
import ImportRemoteLabbookMutation from 'Mutations/ImportRemoteLabbookMutation'
import BuildImageMutation from 'Mutations/BuildImageMutation'
//queries
import UserIdentity from 'JS/Auth/UserIdentity'
//store
import store from 'JS/redux/store'

export default class CreateLabbook extends React.Component {
  constructor(props){

  	super(props);

  	this.state = {
      'name': '',
      'description': '',
      'showError': false,
      'errorType': '',
      'remoteURL': '',
      'textWarning': 'CreateLabbook__warning--hidden',
      'textLength': 0,
      'isUserValid': false,
      'showLoginPrompt': false
    };


    this.continueSave = this.continueSave.bind(this)
    this._updateTextState = this._updateTextState.bind(this)
    this._updateRemoteUrl =  this._updateRemoteUrl.bind(this)
    this._closeLoginPromptModal = this._closeLoginPromptModal.bind(this)

  }

  /**
  *   @param {Object} evt
  *   takes an event input
  *   creates a labbook mutation sets labbook name on parent component
  *   triggers setComponent to proceed to next view
  **/
  continueSave = (evt) => {
    const {name, description} = this.state;
    const id = uuidv4()

    let self = this;

    if(this.state.remoteURL.length > 0){
      const labbookName = this.state.remoteURL.split('/')[this.state.remoteURL.split('/').length - 1]
      const owner = this.state.remoteURL.split('/')[this.state.remoteURL.split('/').length - 2]
      const remote = this.state.remoteURL.indexOf('https://') > -1 ? this.state.remoteURL + '.git' : 'https://' + this.state.remoteURL + '.git'

      UserIdentity.getUserIdentity().then(response => {
        if(navigator.onLine){

          if(response.data){

            if(response.data.userIdentity.isSessionValid){

              store.dispatch(
                {
                  type: "MULTIPART_INFO_MESSAGE",
                  payload: {
                    id: id,
                    message: 'Importing Project please wait',
                    isLast: false,
                    error: false
                  }
                })
              ImportRemoteLabbookMutation(
                owner,
                labbookName,
                remote,
                (response, error) => {


                  if(error){
                    console.error(error)
                    store.dispatch(
                      {
                        type: 'MULTIPART_INFO_MESSAGE',
                        payload: {
                          id: id,
                          message: 'ERROR: Could not import remote Project',
                          messageBody: error,
                          error: true
                      }
                    })

                  }else if(response){

                    store.dispatch(
                      {
                        type: 'MULTIPART_INFO_MESSAGE',
                        payload: {
                          id: id,
                          message: `Successfully imported remote Project ${labbookName}`,
                          isLast: true,
                          error: false
                        }
                      })

                    const labbookName = response.importRemoteLabbook.newLabbookEdge.node.name
                    const owner = response.importRemoteLabbook.newLabbookEdge.node.owner

                    BuildImageMutation(
                    labbookName,
                    owner,
                    false,
                    (response, error)=>{
                      if(error){
                        console.error(error)
                        store.dispatch(
                          {
                            type: 'MULTIPART_INFO_MESSAGE',
                            payload: {
                              id: id,
                              message: `ERROR: Failed to build ${labbookName}`,
                              messsagesList: error,
                              error: true
                          }
                        })
                      }
                    })
                    self.props.history.replace(`/projects/${owner}/${labbookName}`)
                  }else{

                    BuildImageMutation(
                    labbookName,
                    localStorage.getItem('username'),
                    false,
                    (response, error)=>{
                      if(error){
                        console.error(error)
                        store.dispatch(
                          {
                            type: 'MULTIPART_INFO_MESSAGE',
                            payload: {
                              id: id,
                              message: `ERROR: Failed to build ${labbookName}`,
                              messsagesList: error,
                              error: true
                          }
                        })
                      }
                    })
                  }
                }
              )
            }else{
              this.props.auth.renewToken(true, ()=>{
                store.dispatch(
                  {
                    type: "MULTIPART_INFO_MESSAGE",
                    payload: {
                      id: id,
                      message: 'ERROR: User session not valid for remote import',
                      messsagesList: [{message:'User must be authenticated to perform this action.'}],
                      error: true
                    }
                  })

                this.setState({'showLoginPrompt': true})
              }, ()=>{
                this.continueSave();
              });
            }
          }
        } else{
          this.setState({'showLoginPrompt': true})

        }
      })
    }
    else{
      this.props.createLabbookCallback(name, description)
    }
  }
  /**
  *  @param {}
  *  closes login prompt modal
  *  @return {}
  */
  _closeLoginPromptModal(){
    this.setState({
      'showLoginPrompt': false
    })
  }
  /**
    @param {Object, string} evt,field
    updates text in a state object and passes object to setState method
  */
  _updateTextState = (evt, field) =>{
    let state = {}

    state[field] = evt.target.value;
    if(field === 'name'){
      let isMatch = validation.labbookName(evt.target.value)
      this.setState({
      'showError': ((isMatch === false) && (evt.target.value.length > 0)),
      'errorType': ''
      })

      this.props.toggleDisabledContinue((evt.target.value === "") || (isMatch === false));

    }
    let textLength = 260 - evt.target.value.length
    if(textLength >= 100){
      state['textWarning'] = 'CreateLabbook__warning--hidden'
    }else if((textLength <= 100) && (textLength > 50)){
      state['textWarning'] = 'CreateLabbook__warning--green'
    }else if((textLength <= 50) && (textLength > 20)){
      state['textWarning'] = 'CreateLabbook__warning--yellow'
    }else{
      state['textWarning'] = 'CreateLabbook__warning--red'
    }
    state['textLength'] = textLength;

    this.setState(state)
  }


  /**
    @param {}
    returns error message
    @return {string} errorMessage
  */
  _getErrorText(){
    return this.state.errorType === 'send' ? 'Error: Last character cannot be a hyphen.' : 'Error: Title may only contain lowercase alphanumeric and `-`. (e.g. lab-book-title)'
  }

  /**
    @param {event} evt
    updates remote url state
  */
  _updateRemoteUrl(evt){

    this.setState({
      remoteURL: evt.target.value
    })
    if(evt.target.value.length > 0){
      this.props.toggleDisabledContinue(false);
    }else{
      this.props.toggleDisabledContinue(true);
    }
  }

  render(){
    return(
      <div className="CreateLabbook">
        {
          this.state.showLoginPrompt &&
          <LoginPrompt closeModal={this._closeLoginPromptModal}/>
        }
          <div>
            <div>
              <label>Title</label>
              <input
                type='text'
                maxLength="36"
                className={this.state.showError ? 'invalid' : ''}
                onChange={(evt) => this._updateTextState(evt, 'name')}
                placeholder="Enter a unique, descriptive title"
              />
              <span className={this.state.showError ? 'error': 'hidden'}>{this._getErrorText()}</span>
            </div>

            <div>
              <label>Description</label>
              <textarea
                maxLength="260"
                className="CreateLabbook__description-input"
                type="text"
                onChange={(evt) => this._updateTextState(evt, 'description')}

                placeholder="Briefly describe this Project, its purpose and any other key details. "
              />
              <p className={'CreateLabbook__warning ' + this.state.textWarning}>{`${this.state.textLength} characters remaining`}</p>
            </div>

          </div>
        </div>
      )
  }
}
