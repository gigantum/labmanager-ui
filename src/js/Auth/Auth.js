import history from 'JS/history';
import auth0 from 'auth0-js';
import { AUTH_CONFIG } from './auth0-variables';
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation'
//store
import store from 'JS/redux/store'


export default class Auth {
  auth0 = new auth0.WebAuth({
    domain: AUTH_CONFIG.domain,
    clientID: AUTH_CONFIG.clientId,
    redirectUri: AUTH_CONFIG.callbackUrl,
    audience:  AUTH_CONFIG.audience,
    responseType: 'token id_token',
    scope: 'openid profile email user_metadata'
  });

  constructor() {
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);

  }

  login() {
    this.auth0.authorize();
  }

  handleAuthentication() {

    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {

        this.setSession(authResult);

      } else if (err) {
        console.error(err);
      //  alert(`Error: ${err.error}. Check the console for further details.`);
      }
    });
  }

  setSession(authResult) {

    // Set the time that the access token will expire at
    let expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
    localStorage.setItem('access_token', authResult.accessToken);
    localStorage.setItem('id_token', authResult.idToken);
    localStorage.setItem('expires_at', expiresAt);
    localStorage.setItem('family_name', authResult.idTokenPayload.family_name);
    localStorage.setItem('given_name', authResult.idTokenPayload.given_name);
    localStorage.setItem('email', authResult.idTokenPayload.email);
    localStorage.setItem('username', authResult.idTokenPayload.nickname);
    //redirect to labbooks when user logs in

    let storrageRoute = sessionStorage.getItem('CALLBACK_ROUTE')
    let route = storrageRoute !== '' ? storrageRoute : `/labbooks`

    history.replace(route)
  }

  logout() {

    localStorage.removeItem('access_token');
    localStorage.removeItem('id_token');
    localStorage.removeItem('expires_at');
    localStorage.removeItem('family_name')
    localStorage.removeItem('given_name');
    localStorage.removeItem('email');
    localStorage.removeItem('username');
    RemoveUserIdentityMutation(()=>{
      //redirect to root when user logs out
      history.replace('/');
    })
  }

  isAuthenticated() {
    // Check whether the current time is past the
    // access token's expiry time
    let expiresAt = JSON.parse(localStorage.getItem('expires_at'));

    return new Date().getTime() < expiresAt;
  }
}
