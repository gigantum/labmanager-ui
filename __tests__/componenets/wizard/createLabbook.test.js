import React from 'react';
import CreateLabbook from 'Components/wizard/CreateLabbook';
import {mount} from 'enzyme'
import renderer from 'react-test-renderer';
import Auth from 'JS/Auth/Auth';

let toggleDisabledContinue = () => {}
let setComponent = () => {}
let setLabbookName = () => {}

let baseImage = {}

test('Test CreateLabbook rendering', () => {

  const component = renderer.create(
      <CreateLabbook
      toggleDisabledContinue={toggleDisabledContinue}
      setComponent={setComponent}
      setLabbookName={setLabbookName}
      nextWindow={'selectBaseImage'}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});
