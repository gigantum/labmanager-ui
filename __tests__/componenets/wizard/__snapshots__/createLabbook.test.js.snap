// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CreateLabbook rendering 1`] = `
<div
  className="CreateLabbook"
>
  <div
    className="CreateLabbook__modal-inner-container flex flex--column justify--space-between"
  >
    <div>
      <label>
        Title
      </label>
      <input
        className=""
        onChange={[Function]}
        placeholder="Enter a unique, descriptive title"
        type="text"
      />
      <span
        className="hidden"
      >
        Error: Title may only contain alphanumeric characters separated by hyphens. (e.g. lab-book-title)
      </span>
    </div>
    <div>
      <label>
        Description
      </label>
      <textarea
        className="CreateLabbook__description-input"
        onChange={[Function]}
        placeholder="Briefly describe this lab book, its purpose and any other key details. "
        type="text"
      />
    </div>
    <div
      className="CreateLabbook__text-divider-container"
    >
      <span
        className="CreateLabbook__text-divider"
      >
        or
      </span>
    </div>
    <div>
      <label>
        Add public Lab Books
      </label>
      <input
        placeholder="Enter URL Location"
        type="text"
      />
    </div>
  </div>
</div>
`;
